README.md

## About Wind Waker Randomizer

This is a randomizer for The Legend of Zelda: The Wind Waker.  
It randomizes all the items in the game so that each playthrough is unique and you never know where a particular item will be.  
It also makes the game completely open world from the start, removes most cutscenes from the game, and increases sailing speed and text speed.

## Information

The randomizer only supports the North American GameCube version of Wind Waker. (MD5: d8e4d45af2032a081a0f446384e9261b)  
The European and Japanese versions of Wind Waker won't work, and neither will Wind Waker HD.

The randomizer guarantees that every playthrough will be completable, and that you don't need to use any glitches or tricks to beat it.

All items are randomized, but because Wind Waker is such a large game, a single run of it would take a very long time if you had to check every single location. Therefore the randomizer has options to limit where progress items can appear based on the type of the location.  
For example, you can limit progress items to appearing in dungeons and secret caves only, or secret caves sidequests and mail, or any other combination you want.  
Location types that you don't select will only have unimportant items that you don't need to beat the game - like rupees, heart pieces, bomb bag upgrades, etc. So you can skip checking them entirely, unless you want some of those optional items.

If you seem to be stuck and can't find anywhere to progress, you should first consult the spoiler log. The spoiler log is generated at the same time as the randomized ISO, and is put in the same folder. It contains information on everything that was randomized in this seed, and lists the order you can get progress items in as well.

If you've consulted the spoiler log and you're still stuck, it's possible you've encountered a bug in the randomizer.  
Please report bugs like that in the Multiplayer Discord - https://discord.gg/44jYHA7zSA
In the bug report be sure to include the permalink for the seed you encountered the bug on.

If you're going to play on emulator, you should use the latest development version of Dolphin which can be found at the top of this page: https://dolphin-emu.org/download/  
Note that the GameCube boot up animation in Dolphin doesn't work with the randomizer and will cause the game to crash before reaching the main menu. If you have previously set Dolphin up to play that animation you will need to disable it before launching the randomized game by going to Config -> GameCube in Dolphin and checking "Skip Main Menu".  

## Discord Server
This version of the randomizer is a specific Fork of the main Randomizer. For questions related to the Multiplayer Experience, feel free to join the Discord!
https://discord.gg/44jYHA7zSA

If you have questions about Modding the Wind Waker Randomizer, feel free to join the Main Discord!
https://discord.gg/r2963mt

## Credits

This Fork of the randomizer was created and programmed by CrainWWR, with help from:
LagoLunatic (Official Randomizer Creator, this is only possible because of them)
CryZe (event flag documentation)  
Gamma / SageOfMirrors (custom model conversion, file format documentation)  
Hypatia (textures)  
JarheadHME (additional programming)  
LordNed (file format documentation)  
MelonSpeedruns (game design suggestions, graphic design)  
nbouteme (additional programming)  
TrogWW (additional programming)  
wooferzfg (additional programming)
Gymnast86 (New Randomization Logic Base)
Tanjo3 (Hints)

## Running the randomizer from source

If you want to run the latest development/beta version of the randomizer from source, follow the instructions below.

Download and install git from here: https://git-scm.com/downloads  
Then clone this repository with git by running this in a command prompt:  
`git clone https://github.com/LagoLunatic/wwrando.git`  

Download and install Python 3.9.6 from here: https://www.python.org/downloads/release/python-396/  
"Windows x86-64 executable installer" is the one you want if you're on Windows, "macOS 64-bit installer" if you're on Mac.  
If you're on Linux, run this command instead: `sudo apt-get install python3.9`  

Open the wwrando folder in a command prompt and install dependencies by running:  
`py -3.9 -m pip install -r requirements.txt` (on Windows)  
`python3 -m pip install -r requirements.txt` (on Mac)  
`python3 -m pip install $(cat requirements.txt) --user` (on Linux)  

Then run the randomizer with:  
`py -3.9 wwrando.py` (on Windows)  
`python3 wwrando.py` (on Mac)  
`python3 wwrando.py` (on Linux)  

Optionally, you can also install `requirements_full.txt` with the same process you used for `requirements.txt` above.  
`requirements_full.txt` will install additional libraries that speed up texture recoloring, as well as for building a distributable version of the randomizer. You can still run the randomizer from source without these.  


CHANGELOG.md

### Next version (in progress, not released yet)

New features:
* The randomizer now officially supports Mac and Linux and offers release builds for those platforms.
* Added a button to the top of the Player Customization tab to automatically install a custom player model or pack of multiple custom player models from a .zip file.
* Added an optional tweak to invert the X axis of the compass on the sea so that it works more intuitively.
* Forest Haven's shore now has a warp which acts as a shortcut allowing you to quickly re-enter the dungeon entrance from the sea after you've already reached it the normal way once.
* Custom model authors can now change the colors of the sword's slash trail when parrying and when under the Elixir Soup effect.
* Added a command line argument `-autoseed` that generates a random seed name instead of loading the last used one from settings.txt.

Changes:
* The "seed hash" the randomizer adds to the name entry screen is now also included in log files.
* Removed several minor cutscenes from Forest Haven where the camera panned around.
* Inter-dungeon warp pots now have red-colored smoke to differentiate them from other warp pots.
* You must enable Spoiler Log Generation to allow Map Select.

Bug fixes:
* Fixed permalinks not being properly decoded in some cases when Swordless mode is enabled and starting gear is customized.
* Fixed the Song of Passing being logically required for Letter from Baito, even though you do not actually need to wait for time to pass for that letter.
* Fixed the chest the randomizer adds to the DRC miniboss room not appearing on the compass when in a different room.
* Fixed the number of Tingle Statues you own being displayed incorrectly on the Tingle Tuner's GBA screen.

Removed:
* 32-bit builds of the randomizer are no longer offered.

### Version 1.9.0 (released 2021-02-23)

New features:
* When Race Mode is enabled, you can now choose for the number of required dungeons to be anywhere from 1 to 6, instead of only 4 dungeons.
* You can now choose to randomize the Hero's Shield instead of starting the game with it.
* Custom model authors can now change the colors of the sword's slash trail, the boomerang's trail, and arrow trails.
* Slightly increased the speed at which Link sidles along walls.
* The Nintendo Gallery is now unlocked from the start of the game.
* Added a counter showing the number of Tingle Statues you own to the Quest Status screen. This allows you to easily check how many you've picked up without going to Tingle Island. (This replaces the Treasure Chart counter on that screen, but you can still view the number of charts you own in the charts menu.)

Changes:
* Moved the inter-dungeon warp pot in Earth Temple from the fourth room to the second room so that it can be accessed with fewer items than before.
* The option to remove the title/ending videos and the enemy palette randomizer no longer affect the permalink.
* For progress items that have extra duplicates, the randomizer now only guarantees that the required number of them are in progression locations. For example, out of the four Empty Bottles, one will be placed in a progression location if it's needed, but the other three can be placed in either progression or non-progression locations, as if those three were non-progress items.
* Defeating the Mighty Darknuts in the Master Sword chamber with no weapon besides Skull Hammer can no longer be required by the logic.
* Defeating Big Octos that have 12 eyes with arrows can now only be required if you have either the 60 arrow quiver or Light Arrows.
* Renamed "Randomized Sword" to "No Starting Sword" and "Start with Sword" to "Start with Hero's Sword".
* Renamed a few item locations to be more clear.
* The item given for hitting Orca 500 times can now only be consumable items such as rupees (previously it could be anything except progress items).
* The Windfall bomb shop owner now sells bomb refills at reasonable prices from the start of the game, instead of setting the prices to be unaffordable before you own Nayru's Pearl.
* The progress bar now updates incrementally as the randomized ISO is being written.

Bug fixes:
* Fixed a bug where release executables were not properly encrypted, allowing them to be extracted and used to generate spoiler logs for non-spoiler log seeds.
* Fixed a bug where saving the game inside the Pirate Ship on Windfall would cause you to appear on Southern Fairy Island when you reload the game.
* Fixed a bug where certain types of items could be incorrectly counted as progress items regardless of settings if you chose to start with at least 1 of that item but not all of them (e.g. Empty Bottles).
* Fixed a bug where custom player models with tons of color presets built in would cause the randomizer's UI to be very slow when changing custom color options.
* Fixed an error occurring during randomization if your seed name was exactly 42 characters long (the maximum seed length).
* Fixed an error occurring when using a custom ship model that uses the same sail texture as the vanilla ship model.
* Fixed a vanilla bug where mini-boss music would continue to play even after leaving the room if you killed the mini-boss with Light Arrows, or left the room without killing the mini-boss at all.
* Fixed various bugs that could occur when the Fire Mountain and Ice Ring Isle cave entrances were randomized.

### Version 1.8.0 (released 2020-05-15)

New features:
* Implemented color presets for custom player models. The model's author can optionally include sets of colors with the model that can be selected via a dropdown in the randomizer's UI.
  * For example, the vanilla Link model now comes with the following color presets: Dark Link, Smash Dark Link, Green Link, Red Link, Blue Link, Purple Link, Shadow Link.
* Custom player models can now optionally include a custom ship model for the player to ride.
* Custom player models can now allow recoloring the player's sclera.

data/extras.py
import json
import re
from dataclasses import asdict
from collections import OrderedDict

# Replaces the WorldLoadingError Enum Struct in World.hpp
NONE = "NONE"
DUPLICATE_MACRO_NAME = "DUPLICATE_MACRO_NAME"
MACRO_DOES_NOT_EXIST = "MACRO_DOES_NOT_EXIST"
REQUIREMENT_TYPE_DOES_NOT_EXIST = "REQUIREMENT_TYPE_DOES_NOT_EXIST"
MAPPING_MISMATCH = "MAPPING_MISMATCH"
GAME_ITEM_DOES_NOT_EXIST = "GAME_ITEM_DOES_NOT_EXIST"
AREA_DOES_NOT_EXIST = "AREA_DOES_NOT_EXIST"
LOCATION_DOES_NOT_EXIST = "LOCATION_DOES_NOT_EXIST"
EXIT_MISSING_KEY = "EXIT_MISSING_KEY"
OPTION_DOES_NOT_EXIST = "OPTION_DOES_NOT_EXIST"
INCORRECT_ARG_COUNT = "INCORRECT_ARG_COUNT"
EXPECTED_JSON_OBJECT = "EXPECTED_JSON_OBJECT"
AREA_MISSING_KEY = "AREA_MISSING_KEY"
LOCATION_MISSING_KEY = "LOCATION_MISSING_KEY"
MACRO_MISSING_KEY = "MACRO_MISSING_KEY"
REQUIREMENT_MISSING_KEY = "REQUIREMENT_MISSING_KEY"
INVALID_LOCATION_CATEGORY = "INVALID_LOCATION_CATEGORY"
INVALID_MODIFICATION_TYPE = "INVALID_MODIFICATION_TYPE"
INVALID_OFFSET_VALUE = "INVALID_OFFSET_VALUE"
INVALID_GAME_ITEM = "INVALID_GAME_ITEM"

# Replaces the RequirementType Enum in Requirement.hpp
REQUIREMENT_OR = "OR"
REQUIREMENT_AND = "AND"
REQUIREMENT_NOT = "NOT"
REQUIREMENT_HAS_ITEM = "HAS_ITEM"
REQUIREMENT_COUNT = "COUNT"
REQUIREMENT_CAN_ACCESS = "CAN_ACCESS"
REQUIREMENT_SETTING = "SETTING"
REQUIREMENT_MACRO = "MACRO"
REQUIREMENT_NONE = "NONE"

REQUIREMENT_TYPES = [
    REQUIREMENT_OR,
    REQUIREMENT_AND,
    REQUIREMENT_NOT,
    REQUIREMENT_HAS_ITEM,
    REQUIREMENT_COUNT,
    REQUIREMENT_CAN_ACCESS,
    REQUIREMENT_SETTING,
    REQUIREMENT_MACRO,
    REQUIREMENT_NONE
]

# Replaces the SearchMode Enum in Search.hpp
SEARCH_INVALID = "INVALID"
SEARCH_ACCESSIBLE_LOCATIONS = "AccessibleLocations"
SEARCH_GAME_BEATABLE = "GameBeatable"
SEARCH_ALL_LOCATIONS_REACHABLE = "AllLocationsReachable"
SEARCH_GENERATE_PLAYTHROUGH = "GeneratePlaythrough"

LOCATION_CATEGORY_MISC = "Misc"
LOCATION_CATEGORY_DUNGEON = "Dungeon"
LOCATION_CATEGORY_GREAT_FAIRY = "GreatFairy"
LOCATION_CATEGORY_ISLAND_PUZZLE = "IslandPuzzle"
LOCATION_CATEGORY_SPOILS_TRADING = "SpoilsTrading"
LOCATION_CATEGORY_MAIL = "Mail"
LOCATION_CATEGORY_SAVAGE_LABYRINTH = "SavageLabyrinth"
LOCATION_CATEGORY_FREE_GIFT = "FreeGift"
LOCATION_CATEGORY_MINIGAME = "Minigame"
LOCATION_CATEGORY_BATTLE_SQUID = "BattleSquid"
LOCATION_CATEGORY_TINGLE_CHEST = "TingleChest"
LOCATION_CATEGORY_PUZZLE_SECRET_CAVES = "PuzzleSecretCaves"
LOCATION_CATEGORY_COMBAT_SECRET_CAVES = "CombatSecretCaves"
LOCATION_CATEGORY_PLATFORM = "Platform"
LOCATION_CATEGORY_RAFT = "Raft"
LOCATION_CATEGORY_EYE_REEF_CHESTS = "EyeReefChests"
LOCATION_CATEGORY_BIG_OCTO = "BigOcto"
LOCATION_CATEGORY_SUBMARINE = "Submarine"
LOCATION_CATEGORY_GUNBOAT = "Gunboat"
LOCATION_CATEGORY_LONG_SIDE_QUEST = "LongsideQuest"
LOCATION_CATEGORY_SHORT_SIDE_QUESTION = "ShortsideQuest"
LOCATION_CATEGORY_EXPENSIVE_PURCHASES = "ExpensivePurchases"
LOCATION_CATEGORY_SUNKEN_TREASURE = "SunkenTreasure"
LOCATION_CATEGORY_OBSCURE = "Obscure"
LOCATION_CATEGORY_JUNK = "Junk"
LOCATION_CATEGORY_OTHER = "Other"

#Replaces the GameItem map in GameItem.cpp
item_id_dict = {
    "Heart(Pickup)": 0x00,
    "GreenRupee": 0x01,
    "BlueRupee": 0x02,
    "YellowRupee": 0x03,
    "RedRupee": 0x04,
    "PurpleRupee": 0x05,
    "OrangeRupee": 0x06,
    "PieceOfHeart": 0x07,
    "HeartContainer": 0x08,
    "SmallMagicJar(Pickup)": 0x09,
    "LargeMagicJar(Pickup)": 0x0A,
    "5Bombs(Pickup)": 0x0B,
    "10Bombs(Pickup)": 0x0C,
    "20Bombs(Pickup)": 0x0D,
    "30Bombs(Pickup)": 0x0E,
    "SilverRupee": 0x0F,
    "10Arrows(Pickup)": 0x10,
    "20Arrows(Pickup)": 0x11,
    "30Arrows(Pickup)": 0x12,
    "DRCSmallKey": 0x13,
    "DRCBigKey": 0x14,
    "SmallKey": 0x15,
    "Fairy(Pickup)": 0x16,
    "YellowRupee(Joke Message)": 0x1A,
    "DRCDungeonMap": 0x1B,
    "DRCDungeonCompass": 0x1C,
    "FWSmallKey": 0x1D,
    "ThreeHearts(Pickup)": 0x1E,
    "JoyPendant": 0x1F,
    "Telescope": 0x20,
    "TingleTuner": 0x21,
    "WindWaker": 0x22,
    "ProgressivePictoBox": 0x23,
    "SpoilsBag": 0x24,
    "GrapplingHook": 0x25,
    "DeluxePictoBox": 0x26,
    "ProgressiveBow": 0x27,
    "PowerBracelets": 0x28,
    "IronBoots": 0x29,
    "MagicArmor": 0x2A,
    "BaitBag": 0x2C,
    "Boomerang": 0x2D,
    "Hookshot": 0x2F,
    "DeliveryBag": 0x30,
    "Bombs": 0x31,
    "Hero'sClothes": 0x32,
    "SkullHammer": 0x33,
    "DekuLeaf": 0x34,
    "FireandIceArrows": 0x35,
    "LightArrow": 0x36,
    "Hero'sNewClothes": 0x37,
    "ProgressiveSword": 0x38,
    "MasterSword(Powerless)": 0x39,
    "MasterSword(Half Power)": 0x3A,
    "ProgressiveShield": 0x3B,
    "MirrorShield": 0x3C,
    "RecoveredHero'sSword": 0x3D,
    "MasterSword(Full Power)": 0x3E,
    "PieceofHeart(Alternate Message)": 0x3F,
    "FWBigKey": 0x40,
    "FWDungeonMap": 0x41,
    "Pirate'sCharm": 0x42,
    "Hero'sCharm": 0x43,
    "SkullNecklace": 0x45,
    "BokoBabaSeed": 0x46,
    "GoldenFeather": 0x47,
    "Knight'sCrest": 0x48,
    "RedChuJelly": 0x49,
    "GreenChuJelly": 0x4A,
    "BlueChuJelly": 0x4B,
    "DungeonMap": 0x4C,
    "Compass": 0x4D,
    "BigKey": 0x4E,
    "EmptyBottle": 0x50,
    "RedPotion": 0x51,
    "GreenPotion": 0x52,
    "BluePotion": 0x53,
    "ElixirSoup(1/2)": 0x54,
    "ElixirSoup": 0x55,
    "BottledWater": 0x56,
    "FairyinBottle": 0x57,
    "ForestFirefly": 0x58,
    "ForestWater": 0x59,
    "FWDungeonCompass": 0x5A,
    "TotGSmallKey": 0x5B,
    "TotGBigKey": 0x5C,
    "TotGDungeonMap": 0x5D,
    "TotGDungeonCompass": 0x5E,
    "FFDungeonMap": 0x5F,
    "FFDungeonCompass": 0x60,
    "TriforceShard1": 0x61,
    "TriforceShard2": 0x62,
    "TriforceShard3": 0x63,
    "TriforceShard4": 0x64,
    "TriforceShard5": 0x65,
    "TriforceShard6": 0x66,
    "TriforceShard7": 0x67,
    "TriforceShard8": 0x68,
    "NayrusPearl": 0x69,
    "DinsPearl": 0x6A,
    "FaroresPearl": 0x6B,
    "WindsRequiem": 0x6D,
    "BalladOfGales": 0x6E,
    "CommandMelody": 0x6F,
    "EarthGodsLyric": 0x70,
    "WindGodsAria": 0x71,
    "SongOfPassing": 0x72,
    "ETSmallKey": 0x73,
    "ETBigKey": 0x74,
    "ETDungeonMap": 0x75,
    "ETDungeonCompass": 0x76,
    "WTSmallKey": 0x77,
    "BoatsSail": 0x78,
    "Triforce Chart 1 got deciphered": 0x79,
    "Triforce Chart 2 got deciphered": 0x7A,
    "Triforce Chart 3 got deciphered": 0x7B,
    "Triforce Chart 4 got deciphered": 0x7C,
    "Triforce Chart 5 got deciphered": 0x7D,
    "Triforce Chart 6 got deciphered": 0x7E,
    "Triforce Chart 7 got deciphered": 0x7F,
    "Triforce Chart 8 got deciphered": 0x80,
    "WTBigKey": 0x81,
    "All-PurposeBait": 0x82,
    "HyoiPear": 0x83,
    "WTDungeonMap": 0x84,
    "WTDungeonCompass": 0x85,
    "TownFlower": 0x8C,
    "SeaFlower": 0x8D,
    "ExoticFlower": 0x8E,
    "Hero'sFlag": 0x8F,
    "BigCatchFlag": 0x90,
    "BigSaleFlag": 0x91,
    "Pinwheel": 0x92,
    "SickleMoonFlag": 0x93,
    "SkullTowerIdol": 0x94,
    "FountainIdol": 0x95,
    "PostmanStatue": 0x96,
    "ShopGuruStatue": 0x97,
    "Father'sLetter": 0x98,
    "NoteToMom": 0x99,
    "MaggiesLetter": 0x9A,
    "MoblinsLetter": 0x9B,
    "CabanaDeed": 0x9C,
    "ComplimentaryID": 0x9D,
    "Fill-UpCoupon": 0x9E,
    "LegendaryPictograph": 0x9F,
    "DragonTingleStatue": 0xA3,
    "ForbiddenTingleStatue": 0XA4,
    "GoddessTingleStatue": 0XA5,
    "EarthTingleStatue": 0XA6,
    "WindTingleStatue": 0XA7,
    "HurricaneSpin": 0XAA,
    "ProgressiveWallet": 0XAB,
    "5000RupeeWallet": 0XAC,
    "ProgressiveBombBag": 0XAD,
    "99BombBombBag": 0XAE,
    "ProgressiveQuiver": 0XAF,
    "99ArrowQuiver": 0XB0,
    "Nothing": 0XB1,  # THIS REPLACES GameItem::NOTHING
    "MagicMeterUpgrade": 0XB2,
    "50 Rupees, reward for finding 1 Tingle Statue": 0XB3,
    "100 Rupees, reward for finding 2 Tingle Statues": 0XB4,
    "150 Rupees, reward for finding 3 Tingle Statues": 0XB5,
    "200 Rupees, reward for finding 4 Tingle Statues": 0XB6,
    "250 Rupees, reward for finding 5 Tingle Statues": 0XB7,
    "500 Rupees, reward for finding all Tingle Statues": 0XB8,
    "SubmarineChart": 0XC2,
    "Beedle'sChart": 0XC3,
    "PlatformChart": 0XC4,
    "LightRingChart": 0XC5,
    "SecretCaveChart": 0XC6,
    "SeaHeartsChart": 0XC7,
    "IslandHeartsChart": 0XC8,
    "GreatFairyChart": 0XC9,
    "OctoChart": 0XCA,
    "IN-credibleChart": 0XCB,
    "TreasureChart7": 0XCC,
    "TreasureChart27": 0XCD,
    "TreasureChart21": 0XCE,
    "TreasureChart13": 0XCF,
    "TreasureChart32": 0XD0,
    "TreasureChart19": 0XD1,
    "TreasureChart41": 0XD2,
    "TreasureChart26": 0XD3,
    "TreasureChart8": 0XD4,
    "TreasureChart37": 0XD5,
    "TreasureChart25": 0XD6,
    "TreasureChart17": 0XD7,
    "TreasureChart36": 0XD8,
    "TreasureChart22": 0XD9,
    "TreasureChart9": 0XDA,
    "GhostShipChart": 0XDB,
    "Tingle'sChart": 0XDC,
    "TreasureChart14": 0XDD,
    "TreasureChart10": 0XDE,
    "TreasureChart40": 0XDF,
    "TreasureChart3": 0XE0,
    "TreasureChart4": 0XE1,
    "TreasureChart28": 0XE2,
    "TreasureChart16": 0XE3,
    "TreasureChart18": 0XE4,
    "TreasureChart34": 0XE5,
    "TreasureChart29": 0XE6,
    "TreasureChart1": 0XE7,
    "TreasureChart35": 0XE8,
    "TreasureChart12": 0XE9,
    "TreasureChart6": 0XEA,
    "TreasureChart24": 0XEB,
    "TreasureChart39": 0XEC,
    "TreasureChart38": 0XED,
    "TreasureChart2": 0XEE,
    "TreasureChart33": 0XEF,
    "TreasureChart31": 0XF0,
    "TreasureChart23": 0XF1,
    "TreasureChart5": 0XF2,
    "TreasureChart20": 0XF3,
    "TreasureChart30": 0XF4,
    "TreasureChart15": 0XF5,
    "TreasureChart11": 0XF6,
    "TriforceChart8": 0XF7,
    "TriforceChart7": 0XF8,
    "TriforceChart6": 0XF9,
    "TriforceChart5": 0XFA,
    "TriforceChart4": 0XFB,
    "TriforceChart3": 0XFC,
    "TriforceChart2": 0XFD,
    "TriforceChart1": 0XFE,
    "GameBeatable": 0xFF
}

item_id_to_name_dict = {val: key for key, val in item_id_dict.items()}


DUNGEON_NAMES = [
    "DragonRoostCavern",
    "ForbiddenWoods",
    "TowerOfTheGods",
    "ForsakenFortress",
    "EarthTemple",
    "WindTemple",
]
DUNGEON_NAME_DICT = OrderedDict([
    ("DRC", "Dragon Roost Cavern"),
    ("FW", "Forbidden Woods"),
    ("TotG", "Tower of the Gods"),
    ("FF", "Forsaken Fortress"),
    ("ET", "Earth Temple"),
    ("WT", "Wind Temple"),
])

chart_macro_to_island = {
    "Chart for Island 1": "Forsaken Fortress Sunken Treasure",
    "Chart for Island 2": "Star Island Sunken Treasure",
    "Chart for Island 3": "Northern Fairy Sunken Treasure",
    "Chart for Island 4": "Gale Isle Sunken Treasure",
    "Chart for Island 5": "Crescent Moon Sunken Treasure",
    "Chart for Island 6": "Seven Star Sunken Treasure",
    "Chart for Island 7": "Overlook Sunken Treasure",
    "Chart for Island 8": "Four Eye Reef Sunken Treasure",
    "Chart for Island 9": "Mother And Child Sunken Treasure",
    "Chart for Island 10": "Spectacle Sunken Treasure",
    "Chart for Island 11": "Windfall Sunken Treasure",
    "Chart for Island 12": "Pawprint Sunken Treasure",
    "Chart for Island 13": "Dragoon Roost Island Sunken Treasure",
    "Chart for Island 14": "Flight Control Sunken Treasure",
    "Chart for Island 15": "Western Fairy Sunken Treasure",
    "Chart for Island 16": "Rock Spire Sunken Treasure",
    "Chart for Island 17": "Tingle Island Sunken Treasure",
    "Chart for Island 18": "Northern Triangle Sunken Treasure",
    "Chart for Island 19": "Eastern Fairy Sunken Treasure",
    "Chart for Island 20": "Fire Mountain Sunken Treasure",
    "Chart for Island 21": "Star Belt Archipelago Sunken Treasure",
    "Chart for Island 22": "Three Eye Reef Sunken Treasure",
    "Chart for Island 23": "Greatfish Sunken Treasure",
    "Chart for Island 24": "Cyclops Reef Sunken Treasure",
    "Chart for Island 25": "Six Eye Reef Sunken Treasure",
    "Chart for Island 26": "Tower Of The Gods Sunken Treasure",
    "Chart for Island 27": "Eastern Triangle Sunken Treasure",
    "Chart for Island 28": "Thorned Fairy Sunken Treasure",
    "Chart for Island 29": "Needle Rock Sunken Treasure",
    "Chart for Island 30": "Islet Of Steel Sunken Treasure",
    "Chart for Island 31": "Stone Watcher Sunken Treasure",
    "Chart for Island 32": "Southern Triangle Sunken Treasure",
    "Chart for Island 33": "Private Oasis Sunken Treasure",
    "Chart for Island 34": "Bomb Island Sunken Treasure",
    "Chart for Island 35": "Birds Peak Sunken Treasure",
    "Chart for Island 36": "Diamond Steppe Sunken Treasure",
    "Chart for Island 37": "Five Eye Reef Sunken Treasure",
    "Chart for Island 38": "Shark Island Sunken Treasure",
    "Chart for Island 39": "Southern Fairy Sunken Treasure",
    "Chart for Island 40": "Ice Ring Sunken Treasure",
    "Chart for Island 41": "Forest Haven Sunken Treasure",
    "Chart for Island 42": "Cliff Plateau Sunken Treasure",
    "Chart for Island 43": "Horseshoe Sunken Treasure",
    "Chart for Island 44": "Outset Sunken Treasure",
    "Chart for Island 45": "Headstone Sunken Treasure",
    "Chart for Island 46": "Two Eye Reef Sunken Treasure",
    "Chart for Island 47": "Angular Sunken Treasure",
    "Chart for Island 48": "Boating Course Sunken Treasure",
    "Chart for Island 49": "Five Star Sunken Treasure"
}

def handle_lists(obj: list):
    result = []
    for thing in obj:
        if isinstance(thing, list):
            result.append(handle_lists(thing))
        else:
            result.append(asdict(thing))
    return result

def dump_object(obj, filename: str):
    if isinstance(obj, list):
        inp = handle_lists(obj)
    else:
        inp = asdict(obj)
    with open("./dump/" + filename + ".json", 'w') as output:
        json.dump(inp, output, indent='\t')

def capital_case_with_space(original_str: str) -> str:
    pattern = re.compile("[A-Z]([a-z]*)")
    matches = list(map((lambda m: m.group()), pattern.finditer(original_str)))
    if len(matches) > 0:
        return " ".join(matches)
    else:
        return original_str

def dump_simple_world_locations(worlds, output_filename: str) -> None:
    output = list()
    if isinstance(worlds, list):
        for world in worlds:
            output.extend(simple_loc_parse(world))
    else:
        output = simple_loc_parse(worlds)
    with open("./dump/" + output_filename + ".json", 'w') as output_file:
        json.dump(output, output_file, indent='\t')

def simple_loc_parse(world):
    return [location.json_output() for location in world.location_entries]